Improving node connection stability. Adding delay and blocking

fraybyl1 opened the pull request:
potential fix  [#1609](https://github.com/Gozargah/Marzban/issues/1609#issue-2792848603). _maybe not, need testing_

With the current implementation, the time to wait for a response from a node is set to 3 seconds, which probably leads to errors ( e.g. **Read timed out** ) and loss of connection.  The commit makes the following improvements to improve connectivity stability:

Added a delayed retry mechanism to account for short-term network failures. _It may be worth increasing the delay slightly_.

Implemented `threading.Lock` synchronisation in the `get_logs` method to prevent race conditions when accessed simultaneously from multiple threads.

**Again not tested**, please check carefully. After careful review, I see no problems.

Avaritia-ZOV committed:
Improving node connection stability: adding delay and blocking

With the current implementation, the time to wait for a response from a node is set to 3 seconds, which leads to errors ( e.g. Read timed out ) and loss of connection.  The commit makes the following improvements to improve connectivity stability:

Added a delayed retry mechanism to account for short-term network failures. It may be worth increasing the delay slightly.

Implemented threading.Lock synchronisation in the get_logs method to prevent race conditions when accessed simultaneously from multiple threads.

ghost commented:
Looking through the code, I found quite a few potential race conditions. I won't be able to do it today. I hope to get to PR in the next few days.

Avaritia-ZOV committed:
fix: resolve race conditions in node connections

Fixes race states when nodes are connected.

Avaritia-ZOV committed:
refactor: retry logic and error handling

Avaritia-ZOV committed:
fix

Avaritia-ZOV committed:
Once again, refactoring. 

It seems now, I'm confident that everything is working correctly.  Let's go over the changes again compared to the dev branch.

- Moved temporary file management to ManagedTempFile - now files are closed and deleted correctly!
- Improved log collection: added locks and events to prevent threads from jittering.
- Simplified state checking (the started property is now a flag, not an extra query).
- Optimised exception and connection handling to make the API more stable.
- Fixed RPyC service: now connection and logging logic is cleaner (although we don't use it usually, but let it be).

Avaritia-ZOV committed:
A small final fix for node.py

Avaritia-ZOV committed:
feat: Improve node and user management robustness and reliability

Improved user management: the logic of adding, deleting and updating users has been moved to a common function, duplicate code has been removed.

has been moved to a common function, duplicate code has been removed.
In `connect_node` added protection against simultaneous connection of the same node from different threads (blocking and repeated connection attempts).
node from different threads (blocking) and repeated connection attempts.
Improved error handling, added logs. The code became cleaner and clearer.

ghost commented:
I think I'm done with this round of refactoring and optimisations. Together with [#67](https://github.com/Gozargah/Marzban-node/pull/67#issue-2850167282) I think node stability should improve.

ImMohammad20000 commented:
when using rest api for conncting to to node get this error
```
Exception in thread Thread-3 (connect_node):
Traceback (most recent call last):
  File "/.pyenv/versions/3.12.7/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
    self.run()
  File "/.pyenv/versions/3.12.7/lib/python3.12/threading.py", line 1012, in run
    self._target(*self._args, **self._kwargs)
  File "/Marzban/app/xray/operations.py", line 172, in connect_node
    node.start(config)
  File "/Marzban/app/xray/node.py", line 487, in start
    self.remote.start(json_config)
    ^^^^^^^^^^^
  File "/Marzban/app/xray/node.py", line 454, in remote
    self.connect()
  File "/Marzban/app/xray/node.py", line 438, in connect
    raise ConnectionError("Unable to connect to node via RPyC after multiple attempts.")
ConnectionError: Unable to connect to node via RPyC after multiple attempts.

```

Avaritia-ZOV committed:
Fixing REST connectivity.

Connection is possible only via HTTPS. 
Fixed receiving http response from node.

ghost commented:
Should fix the error. If it doesn't. Please show the output: 

`curl -k -v -I -X HEAD --cert /path/to/client_cert.pem --key /path/to/client_key.pem https://<address>:<port>/`

I'll fix it tomorrow.

ImMohammad20000 commented:
Still getting the same error

Avaritia-ZOV committed:
Going back to the past choice of API

ImMohammad20000 commented:
same 

```
Exception in thread Thread-9 (connect_node):
Traceback (most recent call last):
  File "/.pyenv/versions/3.12.7/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
    self.run()
  File "/.pyenv/versions/3.12.7/lib/python3.12/threading.py", line 1012, in run
    self._target(*self._args, **self._kwargs)
  File "/Marzban/app/xray/operations.py", line 172, in connect_node
    node.start(config)
  File "/Marzban/app/xray/node.py", line 487, in start
    self.remote.start(json_config)
    ^^^^^^^^^^^
  File "/Marzban/app/xray/node.py", line 454, in remote
    self.connect()
  File "/Marzban/app/xray/node.py", line 438, in connect
    raise ConnectionError("Unable to connect to node via RPyC after multiple attempts.")
ConnectionError: Unable to connect to node via RPyC after multiple attempts.


```

Avaritia-ZOV committed:
Fixes file handling, improved logging

ImMohammad20000 commented:
when node on rest mode it still try connecting with rpc
```

ERROR:    RPyC connection attempt 1 failed: [Errno 111] Connection refused
ERROR:    RPyC connection attempt 2 failed: [Errno 111] Connection refused
ERROR:    RPyC connection attempt 3 failed: [Errno 111] Connection refused
ERROR:    RPyC connection failed after multiple attempts.
Exception in thread Thread-6 (connect_node):
Traceback (most recent call last):
  File "/home/mohammad/.pyenv/versions/3.12.7/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
    self.run()
  File "/.pyenv/versions/3.12.7/lib/python3.12/threading.py", line 1012, in run
    self._target(*self._args, **self._kwargs)
  File "/Marzban/app/xray/operations.py", line 172, in connect_node
    node.start(config)
  File "/Marzban/app/xray/node.py", line 525, in start
    self.remote.start(json_config)
    ^^^^^^^^^^^
  File "/Marzban/app/xray/node.py", line 497, in remote
    self.connect()
  File "/Marzban/app/xray/node.py", line 479, in connect
    raise ConnectionError(
ConnectionError: Unable to connect to node via RPyC after multiple attempts.

```

Avaritia-ZOV committed:
Small fixes

ImMohammad20000 commented:
error changed
```
/.pyenv/versions/marzban/lib/python3.12/site-packages/urllib3/connection.py:463: SubjectAltNameWarning: Certificate for 172.21.117.3 has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/urllib3/urllib3/issues/497 for details.)
  warnings.warn(
ERROR:    Error making request to /connect: HTTPSConnectionPool(host='172.21.117.3', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname '172.21.117.3' doesn't match 'Gozargah'")))
ERROR:    Connection error (attempt 1): NodeAPIError - (0, 'HTTPSConnectionPool(host=\'172.21.117.3\', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname \'172.21.117.3\' doesn\'t match \'Gozargah\'")))')
/home/mohammad/.pyenv/versions/marzban/lib/python3.12/site-packages/urllib3/connection.py:463: SubjectAltNameWarning: Certificate for 172.21.117.3 has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/urllib3/urllib3/issues/497 for details.)
  warnings.warn(
ERROR:    Error making request to /connect: HTTPSConnectionPool(host='172.21.117.3', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname '172.21.117.3' doesn't match 'Gozargah'")))
ERROR:    Connection error (attempt 2): NodeAPIError - (0, 'HTTPSConnectionPool(host=\'172.21.117.3\', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname \'172.21.117.3\' doesn\'t match \'Gozargah\'")))')
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Stopping main Xray core
INFO:     Stopping nodes Xray core
```

node side error 

```
INFO:     172.21.112.1:12485 - "HEAD / HTTP/1.1" 405 Method Not Allowed
```

Avaritia-ZOV committed:
We fix the problem with ignoring check_hostname=false, with ManagedTempFile working.

ImMohammad20000 commented:
> error changed
> 
> ```
> /.pyenv/versions/marzban/lib/python3.12/site-packages/urllib3/connection.py:463: SubjectAltNameWarning: Certificate for 172.21.117.3 has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/urllib3/urllib3/issues/497 for details.)
>   warnings.warn(
> ERROR:    Error making request to /connect: HTTPSConnectionPool(host='172.21.117.3', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname '172.21.117.3' doesn't match 'Gozargah'")))
> ERROR:    Connection error (attempt 1): NodeAPIError - (0, 'HTTPSConnectionPool(host=\'172.21.117.3\', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname \'172.21.117.3\' doesn\'t match \'Gozargah\'")))')
> /home/mohammad/.pyenv/versions/marzban/lib/python3.12/site-packages/urllib3/connection.py:463: SubjectAltNameWarning: Certificate for 172.21.117.3 has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/urllib3/urllib3/issues/497 for details.)
>   warnings.warn(
> ERROR:    Error making request to /connect: HTTPSConnectionPool(host='172.21.117.3', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname '172.21.117.3' doesn't match 'Gozargah'")))
> ERROR:    Connection error (attempt 2): NodeAPIError - (0, 'HTTPSConnectionPool(host=\'172.21.117.3\', port=62050): Max retries exceeded with url: /connect (Caused by SSLError(CertificateError("hostname \'172.21.117.3\' doesn\'t match \'Gozargah\'")))')
> ^CINFO:     Shutting down
> INFO:     Waiting for application shutdown.
> INFO:     Stopping main Xray core
> INFO:     Stopping nodes Xray core
> ```
> 
> node side error
> 
> ```
> INFO:     172.21.112.1:12485 - "HEAD / HTTP/1.1" 405 Method Not Allowed
> ```

still same error when trying to connect with rest mode
